pkg/proxy/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// used in the middleware below to capture http status code.
+type statusCatcher struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (s *statusCatcher) WriteHeader(statusCode int) {
+	s.statusCode = statusCode
+	s.ResponseWriter.WriteHeader(statusCode)
+}
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		delegate.ServeHTTP(sc, r)
+
+		code := strconv.Itoa(sc.statusCode)
+		verb := strings.ToUpper(r.Method)
+		requestCounter.WithContext(r.Context()).WithLabelValues(verb, code).Inc()
+	})
+}
+
+// WithLatencyTracking tracks the number of seconds it took the wrapped handler
+// to complete.
+func WithLatencyTracking(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		start := time.Now()
+		delegate.ServeHTTP(sc, r)
+		elapsedSeconds := time.Since(start).Seconds()
+
+		code := strconv.Itoa(sc.statusCode)
+		verb := strings.ToUpper(r.Method)
+		requestLatencies.WithContext(r.Context()).WithLabelValues(verb, code).Observe(elapsedSeconds)
+	})
+}
+
+// TODO(csams): enhance these metrics to include shard url
+var (
+	requestCounter = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "proxy_request_total",
+			Help:           "Counter of proxy requests broken out for each verb and HTTP response code.",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"verb", "code"},
+	)
+
+	requestLatencies = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name: "proxy_request_duration_seconds",
+			Help: "Response latency distribution in seconds for each verb and HTTP response code.",
+			// This metric is used for verifying api call latencies SLO,
+			// as well as tracking regressions in this aspects.
+			// Thus we customize buckets significantly, to empower both usecases.
+			Buckets: []float64{0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3,
```

Comment:
Kube provides helpers for these, let's use those if we can. See the authorization cache sync duration metrics we have.

Reply:
Ack - should I use the granularity used in the auth cache or keep what I lifted from https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/metrics/metrics.go#L101-L102?

Reply:
I lifted the wrong ones anyway... :grimacing: I'll change it to use the helper.